### ISSUE TITLE
[Audio] Cannot switch Audio input device (AudioServer.input_device)

### ISSUE THREAD

**Issue by kevinloustau** [godotengine/godot#75603]
### Godot version

v4.0.1.stable.official [cacf49999]

### System information

Windows 11, Sound card zoom h2n, Sound card Roland duo capture

### Issue description

 I try to switch the audio input using the AudioServer.set_input_device() method. Unfortunately, the input device is always the same.
![Screenshot 2023-04-02 181322](https://user-images.githubusercontent.com/4235617/229379356-5d8f10c4-1919-47bc-b848-f58f6d2173a4.png)

### Steps to reproduce

1. Be sure your computer has at leas 2 sound cards (motherboard + external at minimum)
2. Try to use the AudioServer.set_input_device() to change the device

### Minimal reproduction project

[AudioInputDeviceList.zip](https://github.com/godotengine/godot/files/11133376/AudioInputDeviceList.zip)

**Comment by lazydevl0per** [comment]
I have analyzed the problem and found the cause and an initial fix for this problem.
It is also reproducable on the latest commit.

Affected file: drivers\wasapi\audio_driver_wasapi.cpp

Debugging showed that in the current state by using the given project the audio_input is not initialized.
One could initialize an audio_input device in the AudioDriverWASAPI::init() function but i dont know if that is wise regarding privacy? Seems like the way it currently is setup it only initializes an audio_input when input_start() is called. However it is not un-initialized when input_stop() is called so the device still remains as it would in case of using the init() function.. so i presume it would be safe to initialize an audio_input in the init() function first.

After this is done it still wont work since the audio_input.audio_device is never started.
The thread function contains a check if the input is active and only then will it check if the input device has been changed.
Me personally, i think it should be fine to change the input device regardless if its active or not especially AFAIK on windows if there is no microphone permission for the App the input device will stream null.
So we would also need to extract the checks for input device changes out by one scope aswell.


Now things are in a "working" state. I am extremely new to this codebase (just started 2 hours ago) so i am not sure if this would be the "right" solution in case of this engine.

Also since the change of the device seem to be done in a separate thread your print statement in the GoDot script:
`print("Input device: ", AudioServer.input_device)`
might not yet have the updated value of the input device so the print could still output the old one.
I tested this by adding another print before changing and one after and this behaviour seems to be reflected.


Hope i could help

**Comment by kevinloustau** [comment]
I greatly appreciate your help @lazydevl0per 👍

**Comment by lazydevl0per** [comment]
> I greatly appreciate your help @lazydevl0per 👍

Until then you can get the audio device to change if you add an AudioStreamPlayer Node to your scene and choose an AudioStreamPlaybackMicrophone resource as stream source for it.  Make sure to start the AudioStreamPlayer so that the input data is being collected.

**Comment by Calinou** [comment]
Can you reproduce this issue in 3.5.2?

**Comment by lazydevl0per** [comment]
I will try. Need to make a build for 3.5.2 real quick :)

**Comment by MJacred** [comment]
@Calinou, this issue is also present in Godot 3.x: https://github.com/godotengine/godot/issues/46710

**Comment by b4ux1t3** [comment]
Are we sure this is a platform-specific problem? I'm running into it on Linux, using v4.2.1.stable.mono.official [b09f793f5]. 

No matter what I do, I can't get it to switch inputs.

Here's a screenshot from within my editor:

![image](https://github.com/godotengine/godot/assets/5150833/b73d8ec2-2dc4-43ea-8244-de22c956c586)

Here, I'm showing that the input I'm trying to set the server to matches one of the entries in its list of inputs, but it ends up assigning `""` and then `"Default"` instead.

I've tried locking and then unlocking first, no dice. Am I running into a different-yet-similar problem?

**Comment by souplamp** [comment]
> Are we sure this is a platform-specific problem?

Yes, this is specific to Windows because the bug is in AudioDriverWASAPI. Even if the cause is similar, it would require a separate fix than addressing this bug.

**Comment by b4ux1t3** [comment]
> Yes, this is specific to Windows because the bug is in AudioDriverWASAPI. Even if the cause is similar, it would require a separate fix than addressing this bug.
> 
> Open another bug report if you get the chance!

Gotcha! That was basically what I was asking, if I should open another ticket for the Linux issue. 

I'll get the project dusted off, see if recent versions fixed it, and go from there.

Thanks!

**Comment by kevinloustau** [comment]
Hello, any idea about this issue? With the trend of using voice in games, it should very nice to get an eyes on

> > Yes, this is specific to Windows because the bug is in AudioDriverWASAPI. Even if the cause is similar, it would require a separate fix than addressing this bug.
> > Open another bug report if you get the chance!
> 
> Gotcha! That was basically what I was asking, if I should open another ticket for the Linux issue.
> 
> I'll get the project dusted off, see if recent versions fixed it, and go from there.
> 
> Thanks!

Hi @b4ux1t3, Did you have a bit of time to check this issue? 🙂

**Comment by b4ux1t3** [comment]
> Hi @b4ux1t3, Did you have a bit of time to check this issue? 🙂

@kevinloustau 

I have not had time yet, sorry!

I'm planning on doing a round of godot version upgrades with the release in the near-ish future.

**Comment by nalmeth** [comment]
I am also unable to set the AudioServer.output device in 4.3. Giving it anything from the get_output_device_list() just results in it still being 'Default'.

**Comment by kravallee** [comment]
Are there any updates on this? For some Projects using Voice Chat this would be awesome.

**Comment by kevinloustau** [comment]
> Are there any updates on this? For some Projects using Voice Chat this would be awesome.

I agree that with the trend of in game chat audio, this fix would be very welcome 😊

**Comment by lazydevl0per** [comment]
Due to the fact that I keep getting E-mails reminding me of what I started trying to fix here I will setup a Workspace tonight to look at this issue again, seems like 2 years passed the last time I checked.

EDIT:
Managed to get the builds up an running. This bug is still reproducible on 4.5-beta. 
Current progress is here (for now): 
https://github.com/lazydevl0per/godot/tree/bugfix/4.5-beta/75603-WASAPI-Audio-Driver-audio-switching-issue

EDIT2:
I was listening to music on my speakers and caused a funny audio feedback loop and they are busted now heheheh
Anyway, a new PR is open for 4.5-beta, this was fun.

**Comment by goatchurchprime** [comment]
There is a new `AudioServer.get_input_frames(frames)` function in the Godot v4.6 that accesses the microphone directly and potentially avoids all the issues regarding its compatibility with the audio system.

It comes with the function `Error set_input_device_active(active: bool)` so you can turn the input device on and off, instead of this being done implicitly in the background by an `AudioStreamMicrophone` somewhere in the node tree.

I would advise turning off the input_device (ie the microphone) before changing it (by `AudioServer.set_input_device()`) so that you can handle the error case if it fails to turn on.  One of the problems with hot-swapping (changing it when it is active) is there is nowhere to disclose any errors, and I would like to take this feature (of hot-swapping) out.

A demo project which implements the feature for changing the input_device is here: https://github.com/goatchurchprime/godot-demo-projects/tree/gtch/mic_input/audio/mic_input
